Implement layer_shell_v1

OctopusET opened the pull request:
Closes #95

It only works with basic things yet

You can try test with like this
```sh
./cage layer-shell -- -l top # https://gitlab.freedesktop.org/wlroots/wlr-clients/-/blob/master/layer-shell.c
```
or
```sh
./cage wlgreet
```

Hjdskes committed:
Add initial layer shell skeleton

Hjdskes committed:
Add wlr-layer-shell-unstable-v1 protocol

OctopusET commented:
You can try test with like this
`./cage layer-shell -- -l top`
or
`./cage wlgreet`

OctopusET commented:
I will start work again after #313 is merged

septatrix commented:
> I will start work again after #313 is merged

As this is merged, what is the current status of this PR?

OctopusET commented:
I've been busy for a while, I think I will be able to restart work at December

OctopusET commented:
Just rebasing

OctopusET commented:
@septatrix I think now I'm really free. I just started some work. But I will finish other open source work first.

OctopusET commented:
Found a issue why it's crashing
```
[ 517703.732] wl_surface#11.frame(new id wl_callback#14)
[ 517703.763] wl_surface#11.frame(new id wl_callback#19)
[ 517703.819] wl_surface#11.attach(wl_buffer#22, 0, 0)
[ 517703.836] wl_surface#11.damage(0, 0, 2147483647, 2147483647)
[ 517703.857] wl_surface#11.commit()
```
> `wl_surface#11.damage(0, 0, 2147483647, 2147483647)`

figuldan commented:
What is the state of this PR? Any new updates?

SpieringsAE commented:
Would also love to see this for on screen keyboards

OctopusET commented:
Sorry, I've been busy with other things (again...)
I don't think I can work on this in the near future.

OctopusET commented:
I got some free time for some weeks, let me try some.

OctopusET committed:
layer-shell: initial layer shell v1 implementation

OctopusET commented:
Rebased, cleaned more codes.

OctopusET commented:
@emersion would you review this someday when you have time? Thanks

OctopusET committed:
Upgrade wlroots to v0.20

Breaking changes:
 - https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/5230

OctopusET committed:
Drop xdg-shell protocol scanning

wlroots 0.20 generates wayland-protocols headers internally,
so wayland-scanner for xdg-shell is no longer needed.

emersion commented on protocols/meson.build:
```diff
@@ -0,0 +1,37 @@
+  wl_protos_src += wayland_scanner_private_code.process(xml)
+endforeach
+
+lib_server_protos = static_library(
+  'server_protos',
+  wl_protos_headers + wl_protos_src,
+  dependencies: wayland_server.partial_dependency(compile_args: true),
+)
+
+server_protos = declare_dependency(
+  link_with: lib_server_protos,
+  sources: wl_protos_headers,
+)
```
Nit: the old code did this stuff with static libraries and a separate dependency but it was quite unnecessary. I'd recommend using Sway's latest code as a reference:

https://github.com/swaywm/sway/blob/master/protocols/meson.build

(`wl_protos_src` is added to the list of source files specified in `executable()`.)

emersion commented on layer_shell_v1.c:
```diff
@@ -0,0 +1,331 @@
+		if (output_layer) {
+			wlr_scene_node_reparent(&surface->scene->tree->node, output_layer);
+			wlr_scene_node_reparent(&surface->popups->node, &surface->output->server->scene->tree);
+		}
```
Shouldn't we re-arrange the output's layers when the surface is moved between two layers?

emersion commented on layer_shell_v1.c:
```diff
@@ -0,0 +1,331 @@
+		}
+	}
+
+	bool mapped_changed = layer_surface->surface->mapped != surface->mapped;
+	if (!layer_surface->initial_commit && !committed && !mapped_changed) {
+		return;
+	}
+
+	surface->mapped = layer_surface->surface->mapped;
+	arrange_layers(surface->output);
```
Does something break if we remove this code? Doesn't the code in `handle_map()`/`handle_unmap()` already take care of this?

emersion commented on layer_shell_v1.c:
```diff
@@ -0,0 +1,331 @@
+}
+
+static struct cg_layer_popup *
+create_popup(struct wlr_xdg_popup *wlr_popup, struct cg_layer_surface *toplevel, struct wlr_scene_tree *parent)
```
The `toplevel` variable/field name is a tad surprising because usually that refers to `xdg_toplevel`, ie. a regular window. Since we're not dealing with regular windows here, maybe `layer_surface` would be a better name?

emersion commented on layer_shell_v1.c:
```diff
@@ -0,0 +1,331 @@
+	popup->new_popup.notify = popup_handle_new_popup;
+	wl_signal_add(&wlr_popup->base->events.new_popup, &popup->new_popup);
+	popup->commit.notify = popup_handle_commit;
+	wl_signal_add(&wlr_popup->base->surface->events.commit, &popup->commit);
```
Would be nice to install a `reposition` event handler here, so that we support popup repositioning too!

See for instance how Sway does it:

https://github.com/swaywm/sway/blob/909a2ddb5fff528e735341529a028d2ef21836db/sway/desktop/layer_shell.c#L360

emersion commented on output.h:
```diff
@@ -25,5 +36,6 @@ void handle_output_manager_test(struct wl_listener *listener, void *data);
 void handle_output_layout_change(struct wl_listener *listener, void *data);
 void handle_new_output(struct wl_listener *listener, void *data);
 void output_set_window_title(struct cg_output *output, const char *title);
+void arrange_layers(struct cg_output *output);
```
I think we might also want to call this function after an output is reconfigured, e.g. at the end of `handle_new_output()` and in `output_config_apply()` (before updating the output layout, that way views are positioned after).

This can be tested by running e.g. `wlr-randr transform 90` inside cage running with a layer surface opened.

emersion commented on output.c:
```diff
@@ -231,6 +242,18 @@ handle_output_destroy(struct wl_listener *listener, void *data)
+	if (layer == NULL) {
+		return NULL;
+	}
+	layer->node.data = output->wlr_output;
```
Is this actually used for anything?

If we can leave this NULL, then I think we can drop the new view check in `desktop_view_at()`.

emersion commented on seat.c:
I think we might need to revise this comment now.

emersion commented on seat.c:
```diff
@@ -625,8 +631,8 @@ process_cursor_motion(struct cg_seat *seat, uint32_t time_msec, double dx, doubl
 
-	struct cg_view *view = desktop_view_at(seat->server, seat->cursor->x, seat->cursor->y, &surface, &sx, &sy);
-	if (!view) {
+	desktop_view_at(seat->server, seat->cursor->x, seat->cursor->y, &surface, &sx, &sy);
```
I think we need to do something similar for touch input. Without this, touch input will not work on layer surfaces.

emersion commented on seat.c:
```diff
@@ -74,17 +74,23 @@ desktop_view_at(struct cg_server *server, double lx, double ly, struct wlr_surfa
 
-	assert(node != NULL);
-	return node->data;
+	return NULL;
```
If we return NULL here for layer surfaces, that means layer surfaces can't get keyboard focus by clicking them. This is usually what's expected for "on_demand" keyboard interactivity.

Right now `desktop_view_at()` is pretty view-centric, and unfortunately layer surfaces are not views (as in, they are not regular windows). If we wanted to change that we could make it look like this:

```c
bool desktop_at(struct cg_server *server, double lx, double ly, struct wlr_surface **surface, double *sx, double *sy, struct cg_view **view, struct cg_layer_surface **layer_surface);
```

and make it so either `view` or `layer_surface` is populated. Or maybe a better option would be to decouple the function that returns a surface from the one that returns a view or layer surface:

```c
// No longer deals with views
struct wlr_surface *desktop_surface_at(struct cg_server *server, double lx, double ly, struct wlr_surface **surface, double *sx, double *sy);

// Returns a view from any descendant surface (e.g. from main surface, sub-surface, or popup)
struct cg_view *view_from_wlr_surface_descendant(struct wlr_surface *surface);
// Returns a layer surface from any descendant surface (e.g. from main surface, sub-surface, or popup)
struct cg_layer_surface *layer_surface_from_wlr_surface_descendant(struct wlr_surface *surface);
```

The latter two functions are only useful for focus management I believe. Using them would look something like:

```c
struct cg_view *view = view_from_wlr_surface_descendant(surface);
if (view != NULL) {
    seat_set_focus(seat, view);
} else {
    struct cg_layer_surface *layer_surface = layer_surface_from_wlr_surface_descendant(surface);
    if (layer_surface != NULL) {
        seat_set_focus_layer(seat, layer_surface);
    }
}
```

(Perhaps with some more checks based on `keyboard_interactivity`.)

A good layer-shell client to test focus interactions with could be fuzzel configured with `keyboard-focus=on-demand`. It should be possible to switch focus back and forth between fuzzel and a regular window in this mode.

emersion reviewed (commented):
Thanks a lot for working on this patch and keeping it updated, sorry for the wait! This looks quite good to me, here's a first round of comments :)

emersion commented:
>     B: Apply usable_area to view positioning — exclusive_zone shrinks views so bar or popup won't hide main app.

Yeah that sounds definitely like the better option to me!
